### PR TITLE
Feature/crud coins

### LINKED ISSUE #27: Sheet with Money
A sheet has money and there is always 5 types of **CURRENCY** in every DND Sheet: CP (Copper Piece), SP (Silver Piece), EP (Electrum Piece), GP (Gold Piece) and PP (Platinum Piece).

When a Sheet its created they must have them and their **QUANTITY** is initialize as 0 by default. This quantity can be modified by the user but it can't be lower than 0.

In this feature we will implement the following features:
- [Logged users] Create a Sheet with initialized money with the different types of Currency (Like a SheetWallet) 
- [Logged users] Modify the quantity of each SheetCoin when Sheet is updated (Only logged users)
- [Any user] Exchange SheetCoin in differents currencies:

Coins can be exchanged according to the following table of equivalences.

|   Coin   |  Cp  |  Sp  | Ep   | Gp    | Pp     |
|:--------:|:----:|:----:|------|-------|--------|
|  Copper  |   1  | 1/10 | 1/50 | 1/100 | 1/1000 |
|  Silver  |  10  |   1  |  1/5 |  1/10 |  1/100 |
| Electrum |  50  |   5  |   1  |  1/2  |  1/20  |
| Gold     |  100 |  10  |   2  |   1   |  1/10  |
| Platinum | 1000 |  100 |  50  |   10  |    1   |

In this table you can see that 1000 CooperPieces are equal to 1 PlatinumPiece,
1 GoldPiece is equal to 100 CooperPieces and so on.

Example:

If a sheet has 100 Cp, then a user can exchange it into 10 SilverPieces, 2 ElectrumPieces or 1 GoldPiece.
If user tries to exchange that quantity into 1 Platinum it should throw an error such as BadRequest.

### PR DESCRIPTION
**Pull request by zetTtai**
Close #27

### CODE REVIEW

File: DnDCharacterSheet/Models/Coin.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Models
+{
+    public class Coin
+    {
+        [Key]
```
Comment (KiyanDev) [inline DnDCharacterSheet/Models/Coin.cs:8]: i'd avoid having your domain class related to any external representation (think EF entity) by removing this relation you can enjoy a cleaner domain without any references to it's external dependencies (think infrastructure)

File: DnDCharacterSheet/Program.cs [hunk 2]
```diff
@@ -1,22 +1,45 @@
...
+using Microsoft.EntityFrameworkCore;
 using Middlewares;
 using Models;
+using Repositories;
 using Services;
 using Strategies;
 
 var builder = WebApplication.CreateBuilder(args);
 
+var environment = builder.Configuration["Environment"];
```
Comment (KiyanDev) [inline DnDCharacterSheet/Program.cs:15]: This is a nice moment to look into the Options Pattern, you can configure different Configuration Objects ( or just one in this case, something like PersistenceConfiguration ) 

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-8.0

File: DnDCharacterSheet/Repositories/CoinRepository.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            _dbContext.Coins.Add(coin);
+            _dbContext.SaveChanges();
+            return coin;
+        }
+
+        public bool DeleteCoin(long id)
+        {
+            var coinToDelete = GetCoinById(id);
+            if (coinToDelete == null)
```
Comment (KiyanDev) [inline DnDCharacterSheet/Repositories/CoinRepository.cs:21]: Important : Take a look into pattern matching, but my suggestion is to use (T is null / is not null) it has clear advantages and should be preferred when possible, == is an overloadable operator

File: DnDCharacterSheet/Exceptions/NotFoundException.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿namespace Exceptions
+{
+    public class NotFoundException : Exception
```
Comment (KiyanDev) [inline DnDCharacterSheet/Exceptions/NotFoundException.cs:3]: I think this probably can be deleted and use Key not found exception

File: DnDCharacterSheet/DTOs/CoinDTO.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace DTOs
+{
+    public class CoinDTO
+    {
+        public long Id { get; set; }
+        public required string Name { get; set; }
```
Comment (KiyanDev) [inline DnDCharacterSheet/DTOs/CoinDTO.cs:6]: use Data Annotations and keep it consistent

File: DnDCharacterSheet/Converters/CoinConverter.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public class CoinConverter : IConverter<Coin, CoinDTO>
+    {
+        public CoinDTO Convert(Coin source)
+        {
+            return new CoinDTO()
+            {
+                Id = source.Id,
+                Name = source.Name,
+                Initials = source.Initials,
```
Comment (KiyanDev) [inline DnDCharacterSheet/Converters/CoinConverter.cs:15]: see last comma is unnecessary

Comment (KiyanDev) [inline DnDCharacterSheet/Converters/CoinConverter.cs:15]: could it happen Coin to be null? have you setup nullable in the projecT?

File: DnDCharacterSheet/Interfaces/ICoinService.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+{
+    public interface ICoinService
+    {
+        IEnumerable<CoinDTO> GetAllCoins();
+        CoinDTO AddCoin(CoinRequestDTO request);
+        CoinDTO GetCoinById(long id);
+        CoinDTO UpdateCoin(long id, CoinRequestDTO request);
+        bool DeleteCoin(long id);
+    }
+}
```
Comment (KiyanDev) [inline DnDCharacterSheet/Interfaces/ICoinService.cs:13]: Important, i don't think Coin term is correct, what i think you are refering to are Currencies

File: DnDCharacterSheet/Services/CoinService.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        public CoinDTO AddCoin(CoinRequestDTO request)
+        {
+            var coin = new Coin
+            {
+                Name = request.Name ?? throw new BadRequestException(Constants.CoinService.NoNameError),
+                Initials = request.Initials ?? throw new BadRequestException(Constants.CoinService.NoInitialsError),
+            };
+
+            var addedCoin = _repository.AddCoin(coin);
+            return _converter.Convert(addedCoin);
```
Comment (KiyanDev) [inline DnDCharacterSheet/Services/CoinService.cs:35]: These validations could be done in the DTO with Required Attributes so you don't need them here

File: DnDTests/Converters/CoinConverterTest.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            CoinDTO actual = _mapper.Convert(coin);
+
+            // Assert
+            Assert.Multiple(() =>
+            {
+                Assert.That(actual.Id, Is.EqualTo(expected.Id));
+                Assert.That(actual.Name, Is.EqualTo(expected.Name));
+                Assert.That(actual.Initials, Is.EqualTo(expected.Initials));
+            });
+        }
```
Comment (KiyanDev) [inline DnDTests/Converters/CoinConverterTest.cs:46]: use var to allow compiler to infer types 
https://learn.microsoft.com/es-es/dotnet/csharp/programming-guide/classes-and-structs/implicitly-typed-local-variables

**Review by KiyanDev — CHANGES_REQUESTED** [review]

File: DnDCharacterSheet/Converters/CoinConverter.cs (hunk 6 above)
Comment (zetTtai) [reply DnDCharacterSheet/Converters/CoinConverter.cs]: Not possible, and now using Optional Convert is only called when CurrencyRepository successfully returns a Currency.

**Review by KiyanDev — APPROVED** [review]

**Pull request closed without merging**